### PR TITLE
angular2: update with a lot more types and breaking changes

### PR DESCRIPTION
**Pull request by PatrickJS**
the current `angular2.d.ts` is outdated already.
this is the file I built up so far from what I used

### CODE REVIEW

**Comment by vvakame** [comment]
@alexeagle could you review this PR?

File: angular2/angular2.d.ts [hunk 1]
```diff
@@ -12,611 +12,701 @@ interface List<T> extends Array<T> {}
...
-   * <form>
-   *   <input type="text">
-   *   <input type="radio">
-   * <form>
-   * ```
-   *
-   * The directive would only be instantiated on the `<input type="text">` element.
-   *
-   */
   selector: string;
```
Comment (alexeagle) [inline angular2/angular2.d.ts]: Why did you remove the documentation? It's useful in an editor where it's shown side-by-side with the auto-completion.

Comment (gdi2290) [inline angular2/angular2.d.ts]: oh, I didn't know about that

**Comment by alexeagle** [comment]
Note that we are close to publishing the typings directly from the angular2 codebase, so we don't have to hand-maintain this file, and can distribute typings inside our releases.
https://github.com/angular/angular/issues/1966 is the first pass, but I just filed https://github.com/angular/angular/issues/2147 with some follow-up.

This is fine with me in the interim. In particular thanks for fixing the definitions to match name changes since alpha.22!

**Comment by gdi2290** [comment]
yeah np I understand. @vvakame we are good to merge

File: angular2/angular2.d.ts (hunk 1 above)
Comment (vvakame) [inline angular2/angular2.d.ts]: Is this issue resolved?

**Comment by vvakame** [comment]
@alexeagle definition file will removed from DefinitelyTyped repository?

**Comment by alexeagle** [comment]
No, I imagine we'll need two channels for distributing typings:
1) For users who "install" angular, for example with `npm install`, we can include the typings in the distribution
2) For users who just include a `<script>` tag to bootstrap angular, we will still distribute in DefinitelyTyped.

But we need to work out our release process to make that reliable and ensure users have typings that match the binary.

**Comment by vvakame** [comment]
@alexeagle okey, I understand.

**Comment by jcoyne** [comment]
:+1: Thanks. This fixes my problems with current master.

**Comment by alexeagle** [comment]
@vvakame I see you're working on a TS 1.5 branch of DT - how is that going?
We now generate an angular2.d.ts file from Angular's build (sample: https://gist.github.com/alexeagle/c7e0b05927469666d703) but it uses 1.5 syntax so tests fail on master.

It's hopeless to use angular2 with TS 1.4 so I'd rather not spend a lot of time on our .d.ts generator to produce something 1.4 compatible like I did with the original version. Can we check into the 1.5 branch?

But even then, it will be hard for users to fetch it with tsd... thoughts?

**Comment by alexeagle** [comment]
@basarat: any suggestion for my previous comment?

**Comment by basarat** [comment]
> it will be hard for users to fetch it with tsd... thoughts

@vvakame Perhaps `dtsm` can support branches as well? https://github.com/vvakame/dtsm/#reference-other-repository

> so I'd rather not spend a lot of time on our .d.ts generator to produce something 1.4 compatible like I did with the original version

Is that what generated : https://github.com/borisyankov/DefinitelyTyped/pull/4563 ?(which I just merged and is causing the conflict ... so I guess this can be closed)

**Comment by alexeagle** [comment]
@basarat yes this PR can be closed, #4563 should be comprehensive.

I did in fact spend a bunch of time hand-poking 5300 lines to be TS 1.4 compatible, so this will be a problem for our future releases unless we automate the source transformations to lower to TS 1.4.

I'll open a separate issue so we can close this.
